### ISSUE TITLE
document fitting

### ISSUE THREAD

**Issue by hartytp** [OxfordIonTrapGroup/ndscan#204]
The way the fitting is intended to work isn't clear from the existing documentation.

At present there are two main types of analysis: `OnlineFit` and `CustomAnalysis` (inconsistent naming is a little annoying here btw). Some comments:
- ananlyses are only for data fitting, they do not affect plotting; ndscan automatically plots all results channels in the applets (I assume there is some means of modifying how the plotting is done e.g. to use channels as error bars, but I haven't figured that out yet)
- onlinefit is called by the clients (applets etc) each time a data point is pushed to any results channel
- customanalysis is called by the master at the end of the experiment (this is essentially the ndscan replacement for artiq's analyze method)
- a corollary of onlinefit being called by the clients is that the fit results aren't available to the master either to update parameters (e.g. servos) or to save as an experiment result
- customanlaysis results can be broadcast to applets using the annotations (also undocumented at present). locations are for fit results, curve annotations to plot the fits themselves
- customanalysis functions can have result channels passed into them. The idea of this is below (from @dnadlinger on slack):

> Coming back to the analysis channels, the idea for servos is that you'd define the analysis on the bare fragment, so you can play around with it (and have results displayed via the returned plot annotations). Then, to actually use this in an automated servo, you use a TopLevelRunner / subscan / … as required, where you can make use of the numerical return values.
> For a (slightly convoluted, there are multiple axes/…) example, see hoa2_common.highna.piezos
> (There should probably be a convenience function to create a ScanAxis for just one parameter handle – quite often, one doesn't need the capability for * or other wildcard paths.)

- at present if one wants online fits and saved fit results (e.g. for debugging servos that take a while to run) one needs to duplicate the fit code by having both an onlinefit and a customanalysis. This is a bit annoying, particularly as customfit can't take advantage of the boilerplate reduction provided by the online fit interface (e.g. one has to directly call oitg.fitting code). In the future, these should be unified so that the onlinefit is called by the master once all data is collected.

@dnadlinger does that about capture it?

**Comment by hartytp** [comment]
Everything I currently understand about this was gleaned from https://github.com/OxfordIonTrapGroup/ndscan/blob/master/examples/rabi_flop_fit.py which is quite helpful

**Comment by hartytp** [comment]
side note: since OnlineFit is called by the applet, the applet chooses what range to plot over. Currently the applet plots over the full displayed x range, so if the user zooms out then the fit goes to infinity. With the CustomAnalsyis the user has to manually set the fit plot range (without knowing what display range any applet is). This means that once the CustomAnalsysis runs, the plot range will usually change from whatever extrapolation the applet is using for the OnlineFit happens to be using to the CustomAnalysis range (generally the scan limits). Not harmful but just a little surprising.

**Comment by dnadlinger** [comment]
> Not harmful but just a little surprising.

Yep, but I don't see a good way around this – I really want to keep the extended range for online fits, as asymptotic behaviour can be quite useful in verifying whether the fit has actually converged against physical parameters (when e.g. only scanning in a narrow range around some minimum), but for obvious reasons, we can't do that for custom analyses, which can use arbitrary functions for fitting.

There have been some requests for allowing custom OnlineFit functions (#141), the implementation of which (`computed_curve` annotation) could then also be used to display the result of custom fits, but I'm not sure what the sanest way of implementing that is. (Shipping fit functions around the network as Python code?)

**Comment by hartytp** [comment]
I guess the `display_hints` field in setattr_results is what determines how the data is plotted. e.g. 
```python
        self.setattr_result("p_err",
                            min=0.0,
                            display_hints={"error_bar_for": self.p.path})
```

**Comment by hartytp** [comment]
Is it possible to broadcast fit uncertainties using CustomAnalysis location Annotations?

**Comment by dnadlinger** [comment]
> `display_hints`

Yes (see https://oxfordiontrapgroup.github.io/ndscan/apidocs.html#ndscan.experiment.result_channels.ResultChannel), and yes (see `rabi_flop_fit` example – Annotation docs pending…).

**Comment by dnadlinger** [comment]
(Oh, the `rabi_flop_fit` example update from a few days ago isn't pushed yet – apologies!)

**Comment by hartytp** [comment]
Thanks! I'm still having trouble getting my head around things even where the docs are there (it's not obvious what I'm looking for/where to look). Can't tell if the solution would be more high-level docs or just more examples. I'll keep a log of stuff like this and then have a think about the best way to explain it.

**Comment by dnadlinger** [comment]
I think the solution is probably to have more of both – the current state of documentation was never intended to be sufficient without also having our code to go through and learn by example/imitation.

**Comment by dnadlinger** [comment]
`rabi_flop_fit` example with uncertainty region now in master; see https://github.com/OxfordIonTrapGroup/ndscan/blob/f8650f021d03973fd06b1d606ff450db95b0236f/examples/rabi_flop_fit.py#L45.

**Comment by hartytp** [comment]
Yeah, I think a proper manual is in order that builds things up in terms of increasing complexity. I'll have a go at sketching out a few sections at some point in the not too distant future.
